### PR TITLE
bsc light client

### PR DESCRIPTION
**Pull request by xiaoch05**
- Storage proof for precompile contract
- Optimize for storage related to epoch.
- Unit test & benchmark.

### CODE REVIEW

**Comment by HackFisher** [comment]
Remove


```
		/// Verify unsigned relayed headers and finalize authority set
		#[pallet::weight(0)]
		pub fn verify_and_update_authority_set_unsigned(
			origin: OriginFor<T>,
			headers: Vec<BSCHeader>,
		) -> DispatchResultWithPostInfo {
			// ensure not signed
			frame_system::ensure_none(origin)?;

			Self::verify_and_update_authority_set(&headers)?;

			Ok(().into())
		}
```

**Comment by HackFisher** [comment]
Request to simplify the implementation because BSC's epoch length is 200 and block interval is 3 seconds, relaying epoch block is enough for message verification.

Remove
```
pub fn submit_header(
			origin: OriginFor<T>,
			header: BSCHeader,
		) -> DispatchResultWithPostInfo {
...
```

and

```
pub fn verify_header(header: &BSCHeader) -> DispatchResult {
```

**Review by HackFisher — APPROVED** [review]
LGTM, @AurevoirXavier @hujw77 Please help review

File: frame/bridge/bsc/Cargo.toml [hunk 1]
```diff
@@ -11,37 +11,49 @@ version     = "2.7.2"
...
+sp-runtime         = { default-features = false, git = "https://github.com/darwinia-network/substrate", branch = "darwinia-v0.11.8" }
+sp-std             = { default-features = false, git = "https://github.com/darwinia-network/substrate", branch = "darwinia-v0.11.8" }
 
 [dev-dependencies]
 # crates.io
 array-bytes = { version = "1.4" }
-serde_json  = { version = "1.0" }
 # paritytech
 pallet-timestamp = { git = "https://github.com/darwinia-network/substrate", branch = "darwinia-v0.11.8" }
+serde_json       = { version = "1.0" }
```
Comment (AurevoirXavier) [inline frame/bridge/bsc/Cargo.toml:34]: Why move to here?

Comment (xiaoch05) [reply frame/bridge/bsc/Cargo.toml:34]: a mistake

**Review by hujw77 — APPROVED** [review]
LGTM

File: frame/evm/precompile/contracts/bridge/bsc/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+impl<T> Precompile for BscBridge<T>
+where
+	T: darwinia_bridge_bsc::Config,
+{
+	fn execute(
+		input: &[u8],
+		_target_gas: Option<u64>,
+		_context: &Context,
+	) -> core::result::Result<PrecompileOutput, ExitError> {
```
Comment (AurevoirXavier) [inline frame/evm/precompile/contracts/bridge/bsc/src/lib.rs:61]: Why not use `Result` directly

File: frame/evm/precompile/contracts/bridge/bsc/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		let dvm_parser = DvmInputParser::new(&input)?;
+		let (output, cost) = match Action::from_u32(dvm_parser.selector)? {
+			Action::VerfiySingleStorageProof => {
+				let params =
+					BscSingleStorageVerifyParams::decode(dvm_parser.input).map_err(|_| {
+						ExitError::Other("decode single storage verify info failed".into())
```
Comment (AurevoirXavier) [inline frame/evm/precompile/contracts/bridge/bsc/src/lib.rs:67]: Is possible to do

```
map_err(|e| Other(e.to_str()))
```

Comment (xiaoch05) [reply frame/evm/precompile/contracts/bridge/bsc/src/lib.rs:67]: it has no method to do this.

**Comment by AurevoirXavier** [comment]
Let me resolve the conflict.

**Pull request merged**